File: sigv4-signing/go/pkg/sigv4signer/sigv4signer.go
package sigv4signer

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/url"
	"sort"
	"strings"
	"time"
)

type AwsV4Signer struct{}

// SigV4 Signer sample reference for the Kinesis Video Streams WebRTC WebSocket connections
func (signer AwsV4Signer) Sign(uri *url.URL, accessKey string, secretKey string, sessionToken string, region string, dateMilli int64) (*url.URL, error) {

	amzDate := getTimeStamp(dateMilli)
	datestamp := getDateStamp(dateMilli)

	queryParamsMap := buildQueryParamsMap(uri, accessKey, sessionToken, region, amzDate, datestamp)
	canonicalQuerystring := getCanonicalizedQueryString(queryParamsMap)
	canonicalRequest := getCanonicalRequest(uri, canonicalQuerystring)
	stringToSign := signString(amzDate, createCredentialScope(region, datestamp), canonicalRequest)

	signatureKey := getSignatureKey(secretKey, datestamp, region, SERVICE)
	signature := hex.EncodeToString(hmacSha256(stringToSign, signatureKey))

	signedCanonicalQueryString := canonicalQuerystring + "&" + X_AMZ_SIGNATURE + "=" + signature

	signedURI, err := url.Parse(uri.Scheme + "://" + uri.Host + getCanonicalURI(uri) + "?" + signedCanonicalQueryString)
	if err != nil {
		return nil, err
	}

	return signedURI, nil
}

func buildQueryParamsMap(uri *url.URL, accessKey string, sessionToken string, region string, amzDate string, datestamp string) map[string]string {

	xAmzCredential := urlEncode(accessKey + "/" + createCredentialScope(region, datestamp))

	// The SigV4 signer has a maximum time limit of five minutes.
	// Once a connection is established, peers exchange signaling messages,
	// and the P2P connection is successful, the media P2P session
	// can continue for longer period of time.
	queryParamsMap := map[string]string{
		X_AMZ_ALGORITHM:  ALGORITHM_AWS4_HMAC_SHA_256,
		X_AMZ_CREDENTIAL: xAmzCredential,
		X_AMZ_DATE:       amzDate,
		X_AMZ_EXPIRES:    "299",
	}

	if sessionToken != "" {
		queryParamsMap[X_AMZ_SECURITY_TOKEN] = urlEncode(sessionToken)
	}

	queryParamsMap[X_AMZ_SIGNED_HEADERS] = SIGNED_HEADERS

	if uri.RawQuery != "" {
		params := strings.Split(uri.RawQuery, "&")
		for _, param := range params {
			index := strings.Index(param, "=")
			if index > 0 {
				queryParamsMap[param[:index]] = urlEncode(param[index+1:])
			}
		}
	}

	return queryParamsMap
}

func createCredentialScope(region string, datestamp string) string {
	ret := strings.Join([]string{datestamp, region, SERVICE, AWS4_REQUEST_TYPE}, "/")
	return ret
}

func getCanonicalRequest(uri *url.URL, canonicalQuerystring string) string {
	// https://pkg.go.dev/crypto/sha256
	payloadHash := sha256.New().Sum([]byte(""))
	canonicalUri := getCanonicalURI(uri)
	canonicalHeaders := "host:" + uri.Host + NEW_LINE_DELIMITER
	canonicalRequest := strings.Join([]string{METHOD, canonicalUri, canonicalQuerystring, canonicalHeaders, SIGNED_HEADERS, hex.EncodeToString(payloadHash)}, NEW_LINE_DELIMITER)

	return canonicalRequest
}

func getCanonicalURI(uri *url.URL) string {
	if uri.Path == "" {
		return "/"
	}
	return uri.Path
}

func signString(amzDate string, credentialScope string, canonicalRequest string) string {
	// https://pkg.go.dev/crypto/sha256#Sum256
	// Sum256 returns the SHA256 checksum of the data.
	result := sha256.Sum256([]byte(canonicalRequest))
	// Slice to make it to variable size for encoding
	result2 := hex.EncodeToString(result[:])

	stringToSign := strings.Join([]string{ALGORITHM_AWS4_HMAC_SHA_256, amzDate, credentialScope, result2},
		NEW_LINE_DELIMITER)
	return stringToSign
}

func urlEncode(str string) string {
	ret := url.QueryEscape(str)
	return ret
}

func hmacSha256(data string, key []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(data))
	ret := mac.Sum(nil)
	return ret
}

func getSignatureKey(key string, dateStamp string, regionName string, serviceName string) []byte {
	kSecret := []byte("AWS4" + key)
	kDate := hmacSha256(dateStamp, kSecret)
	kRegion := hmacSha256(regionName, kDate)
	kService := hmacSha256(serviceName, kRegion)
	ret := hmacSha256(AWS4_REQUEST_TYPE, kService)
	return ret
}

func getTimeStamp(dateMilli int64) string {
	ret := time.UnixMilli(dateMilli).UTC().Format(TIME_PATTERN)
	return ret
}

func getDateStamp(dateMilli int64) string {
	ret := time.UnixMilli(dateMilli).UTC().Format(DATE_PATTERN)
	return ret
}

func getCanonicalizedQueryString(queryParamsMap map[string]string) string {
	var queryKeys []string
	for key := range queryParamsMap {
		queryKeys = append(queryKeys, key)
	}
	sort.Strings(queryKeys)

	var builder strings.Builder
	for i, key := range queryKeys {
		builder.WriteString(key + "=" + queryParamsMap[key])
		if i < len(queryKeys)-1 {
			builder.WriteString("&")
		}
	}

	return builder.String()
}

File: python/kvs_web_streamer/streaming/sigv4_signer.py
import hashlib
import hmac
from datetime import datetime
from urllib.parse import urlparse

class SigV4Signer:
    """AWS Signature Version 4 signer for KVS"""
    
    def __init__(self, access_key: str, secret_key: str, session_token: str, region: str):
        self.access_key = access_key
        self.secret_key = secret_key
        self.session_token = session_token
        self.region = region
        self.service = 'kinesisvideo'
    
    def sign_request(self, method: str, url: str, headers: dict, payload: bytes) -> dict:
        """Sign HTTP request with SigV4"""
        parsed = urlparse(url)
        uri = parsed.path or '/'
        
        now = datetime.utcnow()
        amz_date = now.strftime('%Y%m%dT%H%M%SZ')
        date_stamp = now.strftime('%Y%m%d')
        
        # Add AWS headers
        headers['X-Amz-Date'] = amz_date
        if self.session_token:
            headers['X-Amz-Security-Token'] = self.session_token
        
        # For POST, use UNSIGNED-PAYLOAD
        if method == 'POST':
            headers['x-amz-content-sha256'] = 'UNSIGNED-PAYLOAD'
            payload_hash = 'UNSIGNED-PAYLOAD'
        else:
            payload_hash = hashlib.sha256(payload).hexdigest()
            headers['x-amz-content-sha256'] = payload_hash
        
        # Canonical request
        canonical_headers = ''.join(f'{k.lower()}:{v.strip()}\n' for k, v in sorted(headers.items()))
        signed_headers = ';'.join(sorted(k.lower() for k in headers.keys()))
        canonical_request = f'{method}\n{uri}\n\n{canonical_headers}\n{signed_headers}\n{payload_hash}'
        
        print(f"[SIGV4] Canonical Request:\n{canonical_request}")
        print(f"[SIGV4] Canonical Request Hash: {hashlib.sha256(canonical_request.encode()).hexdigest()}")
        
        # String to sign
        credential_scope = f'{date_stamp}/{self.region}/{self.service}/aws4_request'
        string_to_sign = f'AWS4-HMAC-SHA256\n{amz_date}\n{credential_scope}\n{hashlib.sha256(canonical_request.encode()).hexdigest()}'
        
        print(f"[SIGV4] String to Sign:\n{string_to_sign}")
        
        # Signature
        k_date = hmac.new(f'AWS4{self.secret_key}'.encode(), date_stamp.encode(), hashlib.sha256).digest()
        k_region = hmac.new(k_date, self.region.encode(), hashlib.sha256).digest()
        k_service = hmac.new(k_region, self.service.encode(), hashlib.sha256).digest()
        k_signing = hmac.new(k_service, b'aws4_request', hashlib.sha256).digest()
        signature = hmac.new(k_signing, string_to_sign.encode(), hashlib.sha256).hexdigest()
        
        print(f"[SIGV4] Signature: {signature}")
        
        # Authorization header
        headers['Authorization'] = f'AWS4-HMAC-SHA256 Credential={self.access_key}/{credential_scope}, SignedHeaders={signed_headers}, Signature={signature}'
        
        return headers
